qa: feature_framework_startup_failures.py fixes & improvements (#30660 follow-up)

hodlinator opened the pull request:
* Handle multiple `--tmpdir` args properly.
* Handle `--timeout-factor=0` properly (fixes #32506).
* Improve readability of expected error message (`assert_raises_message()`).
* Make suppressed error output less confusing (`wait_for_rpc_connection()`).

hodlinator committed:
qa: Fix dormant bug caused by multiple --tmpdir

We would only modify the parent process' first --tmpdir arg.

Now we tack on an additional --tmpdir after the parent's arguments. Also simplifies the code.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

hodlinator committed:
qa: Improve suppressed errors output

Original discussion: https://github.com/bitcoin/bitcoin/pull/30660#discussion_r2080132673

hodlinator committed:
qa: assert_raises_message() - search in str(e)

repr() is annoying because it requires escaping special characters, use str() instead.

Original discussion: https://github.com/bitcoin/bitcoin/pull/30660#discussion_r2080292165

i-am-yuvi commented:
Tested ACK c47f634718d4248fd2a30e51a57944f89da72a64

```
./build/test/functional/test_runner.py feature_framework_startup_failures.py --timeout-factor=0
Temporary test directory at /var/folders/jb/wlbrz0t95vl58wzxjt75wqmh0000gn/T/test_runner_₿_🏃_20250516_012813
Remaining jobs: [feature_framework_startup_failures.py]
1/1 - feature_framework_startup_failures.py passed, Duration: 5 s

TEST                                  | STATUS    | DURATION

feature_framework_startup_failures.py | ✓ Passed  | 5 s

ALL                                   | ✓ Passed  | 5 s (accumulated) 
Runtime: 5 s
```

ismaelsadeeq commented on test/functional/feature_framework_startup_failures.py:
```diff
@@ -42,10 +42,13 @@ def _verify_startup_failure(self, test, internal_args, expected_exception):
         args = [sys.executable] + sys.argv + [f"--tmpdir={self.options.tmpdir}/{subdir}", f"--internal_test={test.__name__}"] + internal_args
 
+        # BitcoinTestFramework.parse_args converts --timeout-factor=0 to 99999.
+        timeout = None if self.options.timeout_factor == 99999 else 60 * self.options.timeout_factor
```
In "qa: Handle --timeout-factor=0" 1235a0df5ff6278d07dc95e7acc3ccb14f536286 whats the rationale behind multiplying by factor of 60?

Wondering if this is the right fix? if I pass timeout-factor=99998 the test will fail, but on master I dont think running the functional test with timeout-factor=99998 will trigger any failure.

hodlinator replied:
> whats the rationale behind multiplying by factor of 60?

Default timeout-factor is 1, so we get 60 seconds.

Here's a similar pre-existing example:

https://github.com/bitcoin/bitcoin/blob/c47f634718d4248fd2a30e51a57944f89da72a64/test/functional/test_framework/test_framework.py#L765-L773

> on master I dont think running the functional test with timeout-factor=99998 will trigger any failure.

99998 does fail feature_framework_startup_failures.py on master, but skipping back before the test was added, it succeeded for me (except `p2p_node_network_limited.py --v1transport, p2p_node_network_limited.py --v2transport` "never" finishes). It appears no other test is specifying the timeout parameter to `subprocess.run()`. It feels more correct to do so, so I'd rather keep it.

> an alternative fix could also be to reduce the magic value from 99'999 to 9'999 or 999. ( Any value of that should be more than sufficient in practise. And anyone really wanting a larger timeout factor, can just set it trivially)

Thanks! Set to 999 in test_framework.py in latest push and reverted the change to feature_framework_startup_failures.py.

ismaelsadeeq commented on test/functional/test_framework/test_node.py:
```diff
@@ -335,16 +339,14 @@ def wait_for_rpc_connection(self, *, wait_for_import=True):
+                latest_error = suppress_error("missing_credentials", e)
             time.sleep(1.0 / poll_per_s)
-        self._raise_assertion_error(f"Unable to connect to bitcoind after {self.rpc_timeout}s (ignored errors: {str(dict(suppressed_errors))}, latest error: {latest_error})")
+        self._raise_assertion_error(f"Unable to connect to bitcoind after {self.rpc_timeout}s (ignored errors: {dict(suppressed_errors)!s}{'' if latest_error is None else f', latest: {latest_error[0]!r}/{latest_error[1]}'})")
```
Nice this is much better I think.

ismaelsadeeq reviewed (commented):
Concept ACK
thanks for following up.

hodlinator committed:
qa: Make --timeout-factor=0 result in a smaller factor

Would otherwise cause an OverflowError in feature_framework_startup_failures.py when calling subprocess.run() with 60 * factor.

Fixes #32506

Co-authored-by: MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>

janb84 reviewed (commented):
LGTM ACK https://github.com/bitcoin/bitcoin/commit/bf950c4544d3a8478b46faf0b93c0dc647274c9b

- code review ✅
- build and tested ✅

Improved code readability and addressed open issues from #30660

achow101 commented:
ACK bf950c4544d3a8478b46faf0b93c0dc647274c9b
